streamer/src/quic.rs

```diff
@@ -0,0 +1,421 @@
+                    let mut packet = Packet::default();
+                    packet.meta.set_addr(remote_addr);
+                    batch.packets.push(packet);
+                    *maybe_batch = Some(batch);
+                }
+
+                if let Some(batch) = maybe_batch.as_mut() {
+                    let end = chunk.offset as usize + chunk.bytes.len();
+                    batch.packets[0].data[chunk.offset as usize..end].copy_from_slice(&chunk.bytes);
+                    batch.packets[0].meta.size = std::cmp::max(batch.packets[0].meta.size, end);
+                }
+            } else {
+                debug!("chunk is none");
+                // done receiving chunks
+                if let Some(batch) = maybe_batch.take() {
+                    let len = batch.packets[0].meta.size;
+                    if let Err(e) = packet_sender.send(batch) {
+                        info!("send error: {}", e);
+                    } else {
+                        trace!("sent {} byte packet", len);
+                    }
+                }
+                return true;
+            }
+        }
+        Err(e) => {
+            debug!("Received stream error: {:?}", e);
+            return true;
+        }
+    }
+    false
+}
+
+pub fn spawn_server(
+    sock: UdpSocket,
+    keypair: &Keypair,
+    gossip_host: IpAddr,
+    packet_sender: Sender<PacketBatch>,
+    exit: Arc<AtomicBool>,
+) -> Result<thread::JoinHandle<()>, QuicServerError> {
+    let (config, _cert) = configure_server(keypair, gossip_host)?;
+
+    let runtime = rt();
+    let (_, mut incoming) = {
+        let _guard = runtime.enter();
+        Endpoint::new(EndpointConfig::default(), Some(config), sock)
+            .map_err(|_e| QuicServerError::EndpointFailed)?
+    };
+
+    let handle = thread::spawn(move || {
+        let handle = runtime.spawn(async move {
+            while !exit.load(Ordering::Relaxed) {
+                const WAIT_FOR_CONNECTION_TIMEOUT_MS: u64 = 1000;
+                let timeout_connection = timeout(
+                    Duration::from_millis(WAIT_FOR_CONNECTION_TIMEOUT_MS),
+                    incoming.next(),
+                )
+                .await;
+
+                if let Ok(Some(connection)) = timeout_connection {
+                    if let Ok(new_connection) = connection.await {
+                        let exit = exit.clone();
+                        let quinn::NewConnection {
+                            connection,
+                            mut uni_streams,
+                            ..
+                        } = new_connection;
+
+                        let remote_addr = connection.remote_address();
+                        let packet_sender = packet_sender.clone();
+                        tokio::spawn(async move {
+                            debug!("new connection {}", remote_addr);
+                            while let Some(Ok(mut stream)) = uni_streams.next().await {
```

Comment:
The stream times itself out after some time, so I don't think it's strictly necessary. I didn't see any problems in tests, but it might be a good idea.